base/applications/rapps/geninst.cpp

```diff
@@ -0,0 +1,811 @@
+{
+    InstallInfo &Info = *static_cast<InstallInfo *>(g_pInfo);
+    UINT csidl = Info.PerUser ? CSIDL_PROGRAMS : CSIDL_COMMON_PROGRAMS;
+    CStringW rel = Info.ShortcutFile, path, dir, tmp;
+
+    if (FAILED(GetSpecialPath(csidl, path, Info.GetGuiOwner())))
+        return TRUE; // Pretend everything is OK
+
+    int cat;
+    if (Info.Parser.GetInt(DB_CATEGORY, cat) && cat == ENUM_CAT_GAMES)
+    {
+        if (GetLocalizedSMFolderName(L"SM_GamesName", L"shortcuts.inf", L"Games", tmp))
+        {
+            path = BuildPath(path, tmp);
+        }
+    }
+
+    rel.Replace('/', '\\');
+    path = BuildPath(path, rel.GetString());
+    UINT SHPPFW = SHPPFW_DIRCREATE | SHPPFW_IGNOREFILENAME;
+    HRESULT hr = SHPathPrepareForWriteW(Info.GetGuiOwner(), NULL, path, SHPPFW);
+    if ((Info.Error = ErrorFromHResult(hr)) != 0)
+    {
+        ErrorBox(Info.Error);
+        return FALSE;
+    }
+
+    CComPtr<IShellLinkW> link;
+    hr = link.CoCreateInstance(CLSID_ShellLink, IID_IShellLinkW);
+    if (SUCCEEDED(hr))
+    {
+        if (SUCCEEDED(hr = link->SetPath(Target)))
+        {
+            if (SUCCEEDED(GetCustomIconPath(Info, tmp)))
+            {
+                LPWSTR p = tmp.GetBuffer();
+                int idx = PathParseIconLocation(p);
+                link->SetIconLocation(p, idx);
+            }
+            CComPtr<IPersistFile> persist;
+            if (SUCCEEDED(hr = link->QueryInterface(IID_IPersistFile, (void**)&persist)))
+            {
+                hr = persist->Save(path, FALSE);
+            }
+        }
+    }
+    if (SUCCEEDED(hr))
+    {
+        if (AddEntry(UNOP_FILE, path))
+        {
+            SplitFileAndDirectory(path, &dir);
+            AddEntry(UNOP_EMPTYDIR, dir);
+        }
+    }
+    else
+    {
+        ErrorBox(ErrorFromHResult(hr));
+    }
+    return !Info.Error;
+}
+
+static BOOL
+InstallFiles(const CStringW &SourceDirBase, const CStringW &Spec,
+             const CStringW &DestinationDir)
+{
+    InstallInfo &Info = *static_cast<InstallInfo *>(g_pInfo);
+    CStringW sourcedir, filespec;
+    filespec = SplitFileAndDirectory(Spec, &sourcedir); // Split "[OptionalDir\]*.ext"
+    sourcedir = BuildPath(SourceDirBase, sourcedir);
+    BOOL failed = FALSE, dots;
+    WIN32_FIND_DATAW wfd;
+    HANDLE hFind = FindFirstFileW(BuildPath(sourcedir, filespec), &wfd);
+    if (hFind == INVALID_HANDLE_VALUE)
+    {
+        DWORD error = GetLastError();
+        if (error == ERROR_FILE_NOT_FOUND)
+            return TRUE;
+        else
+            return !ErrorBox(error);
+    }
+
+    for (;;)
+    {
+        CStringW from = BuildPath(sourcedir, wfd.cFileName);
+        CStringW to = BuildPath(DestinationDir, wfd.cFileName);
+        CStringW uninstpath = to.Mid(Info.InstallDirLen - 1);
+        if (wfd.dwFileAttributes & FILE_ATTRIBUTE_DIRECTORY)
+        {
+            LPWSTR p = wfd.cFileName;
+            dots = (p[0] == '.' && !p[1]) || (p[0] == '.' && p[1] == '.' && !p[2]);
```

Comment:
p[0] check could be factored out.
```suggestion
            BOOL dots = (p[0] == '.') && (!p[1] || (p[1] == '.' && !p[2]));
```